Leak PythonEngine singleton

malfet opened the pull request:
Avoid autograd threads global shutdown logic complexity by leaking PythongEngine and relying on the OS to do resource cleanup

malfet committed:
Leak PythonEngine singleton

Avoid autograd threads global shutdown logic complexity by leaking PythongEngine and relying on the OS to do resource cleanup

albanD reviewed (commented):
nit: I would add a comment in the destructor to mention that we assume there that the python context is still valid (and thus this function should not be called during shutdown).

albanD commented:
Ho we actually have a test that ensures that we did that :o

malfet committed:
Delete TestAutograd.test_thread_shutdown

Also, add a comment to `PythonEngine::thread_init` that global threads are leaked during shutdown

malfet committed:
Revert pybind11 to stock v2.6.0

malfet commented:
Updated PR by deleting the test and reverting pybind to stock 2.6.0

albanD commented on torch/csrc/autograd/python_engine.cpp:
```diff
@@ -33,38 +33,37 @@ namespace torch { namespace autograd { namespace python {
-  if (should_increment) {
-    increment_non_reentrant_thread_count();
-  }
+  //Ignore `should_increment` option, as PythonEngine device threads should never be destroyed
```
I'm not sure about this.
In particular, after a fork, we want to properly clean and reset the threads.
The way it is done is pretend that no thread were actually attached and recreate them.

But if the whole logic to wait for the threads to die and count them is not used anymore, we should remote it right?

malfet replied:
I think it's still valuable to preserve the logic for PyTorch C++ (i.e. base `torch::autograd::Engine`) but kill it for `torch::autograd::python::PythonEngine`

albanD replied:
Ok.
I guess that's another bullet for the "rework autograd engine threading" task :D

albanD reviewed (approved):
lgtm once CI is green
